scripts/build-windows.ps1

```diff
@@ -0,0 +1,175 @@
+    if ( $Env:APPVEYOR_BUILD_WORKER_IMAGE -eq 'Visual Studio 2019' ) { $VS_VERSION='2019' }
+    if ( $Env:APPVEYOR_BUILD_WORKER_IMAGE -eq 'Visual Studio 2015' ) { $VS_VERSION='2015' }
+    if ( $Env:APPVEYOR_BUILD_WORKER_IMAGE -eq 'Visual Studio 2013' ) { $VS_VERSION='2013' }
+
+    #if not statically linking OpenSSL, set flag to gather the specific openssl package from the build server into package
+    if ( $STATIC_LINK_SSL -eq 'OFF' ) { $Env:GATHER_SSL_INTO_PACKAGE = $true }
+
+    #if unit tests are on, set flag to actually execute ctest step
+    if ( $UNIT_TESTS -eq 'ON' ) { $Env:RUN_UNIT_TESTS = $true }
+    
+    $CONFIGURATION = $Env:CONFIGURATION
+
+    #appveyor has many openssl installations - place the latest one in the default location unless VS2013
+    if( $VS_VERSION -ne '2013' ){
+        Remove-Item -Path "C:\OpenSSL-Win32" -Recurse -Force -ErrorAction SilentlyContinue | Out-Null
+        Remove-Item -Path "C:\OpenSSL-Win64" -Recurse -Force -ErrorAction SilentlyContinue | Out-Null
+        Copy-Item -Path "C:\OpenSSL-v111-Win32" "C:\OpenSSL-Win32" -Recurse | Out-Null
+        Copy-Item -Path "C:\OpenSSL-v111-Win64" "C:\OpenSSL-Win64" -Recurse | Out-Null
+    }
+}
+
+# persist VS_VERSION so it can be used in an artifact name later
+$Env:VS_VERSION = $VS_VERSION
+
+# select the appropriate cmake generator string given the environment
+if ( $VS_VERSION -eq '2019' ) { $CMAKE_GENERATOR = 'Visual Studio 16 2019'; $MSBUILDVER = "16.0"; }
+if ( $VS_VERSION -eq '2015' -and $DEVENV_PLATFORM -eq 'Win32' ) { $CMAKE_GENERATOR = 'Visual Studio 14 2015'; $MSBUILDVER = "14.0"; }
+if ( $VS_VERSION -eq '2015' -and $DEVENV_PLATFORM -eq 'x64' ) { $CMAKE_GENERATOR = 'Visual Studio 14 2015 Win64'; $MSBUILDVER = "14.0"; }
+if ( $VS_VERSION -eq '2013' -and $DEVENV_PLATFORM -eq 'Win32' ) { $CMAKE_GENERATOR = 'Visual Studio 12 2013'; $MSBUILDVER = "12.0"; }
+if ( $VS_VERSION -eq '2013' -and $DEVENV_PLATFORM -eq 'x64' ) { $CMAKE_GENERATOR = 'Visual Studio 12 2013 Win64'; $MSBUILDVER = "12.0"; }
+
+# clear any previous build and create & enter the build directory
+$buildDir = Join-Path "$projectRoot" "_build"
+Write-Output "Creating (or cleaning if already existing) the folder $buildDir for project files and outputs"
+Remove-Item -Path $buildDir -Recurse -Force -ErrorAction SilentlyContinue | Out-Null
+New-Item -ItemType Directory -Path $buildDir -ErrorAction SilentlyContinue | Out-Null
+Push-Location $buildDir
+
+# check cmake is installed
+if ($null -eq (Get-Command "cmake.exe" -ErrorAction SilentlyContinue)) 
+{ 
+    $installCmake = Read-Host "Unable to find cmake in your PATH - would you like to download and install automatically? [yes/no]"
+   
+    if ($installCmake -eq "y" -or $installCmake -eq "yes") {
+        # download cmake 3.17.3 64-bit and run MSI for user
+        $client = New-Object System.Net.WebClient        
+        $tempDownloadFile = New-TemporaryFile
+        $cmakeMsiFile = "$tempDownloadFile.cmake-3.17.3-win64-x64.msi"
```

Comment:
I like this approach to offer a user a possibility to install cmake when it's not present - but hardcoding the cmake version here is rather a bad practice. At least allow a user to edit the download path, and return to this step if download failed.

Reply:
Yeah, this was just me jumping between machines when testing and deciding to set this option. I thought about making it 'soft-set', but that would then end up polluting the input parameters to the script with some significant extra noise for a relatively low benefit.

As a compromise, how about the cmake version URL gets promoted to a 'top of script' variable that is easy for people to spot and edit as needed (and will be easier to review any PR for an update)?

Reply:
Yes, good idea, and moreover, maybe a possibility to provide the version itself in an environment variable? I think just the version itself `3.17.3` should suffice to create a long-time-valid (I hope) URI.

Reply:
Environment var seems overkill, since this is really intended for people running interactively (so likely don't have an environment variable set - it's easier to tweak a text file before running than try and set an environment variable. It would also clutter the variable defined in the script and lead to potential confusion if one is set yet this value is override.
If this was being used within the automation - sure; but the MSI just fires to run interactively and then requires the script to run again to have the path discoverable.

So I have just adapted the version number to patch the URL (and added a warning comment that this might not be 100% reliable) and committed - hopefully that's a good middle ground.

Reply:
Ok, we can live with a possible modifying the script to put the right version.